components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,635 @@
+                if from != last_index + 1 {
+                    // Entries are put and deleted in the same writebatch.
+                    return Err(box_err!(
+                        "invalid cut logs, last_index {}, cut_logs {:?}",
+                        last_index,
+                        self.cut_logs
+                    ));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Message that can be sent to Worker
+pub enum WriteMsg<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    WriteTask(WriteTask<EK, ER>),
+    Shutdown,
+}
+
+impl<EK, ER> fmt::Debug for WriteMsg<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            WriteMsg::WriteTask(t) => write!(
+                fmt,
+                "WriteMsg::WriteTask(region_id {} peer_id {} ready_number {})",
+                t.region_id, t.peer_id, t.ready_number
+            ),
+            WriteMsg::Shutdown => write!(fmt, "WriteMsg::Shutdown"),
+        }
+    }
+}
+
+/// WriteTaskBatch is used for combining several WriteTask into one.
+struct WriteTaskBatch<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub kv_wb: EK::WriteBatch,
+    pub raft_wb: ER::LogBatch,
+    // Write raft state once for a region everytime writing to disk
+    pub raft_states: HashMap<u64, RaftLocalState>,
+    pub state_size: usize,
+    pub tasks: Vec<WriteTask<EK, ER>>,
+    // region_id -> (peer_id, ready_number)
+    pub readies: HashMap<u64, (u64, u64)>,
+}
+
+impl<EK, ER> WriteTaskBatch<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    fn new(kv_wb: EK::WriteBatch, raft_wb: ER::LogBatch) -> Self {
+        Self {
+            kv_wb,
+            raft_wb,
+            raft_states: HashMap::default(),
+            state_size: 0,
+            tasks: vec![],
+            readies: HashMap::default(),
+        }
+    }
+
+    /// Add write task to this batch
+    fn add_write_task(&mut self, mut task: WriteTask<EK, ER>) {
+        if let Err(e) = task.valid() {
+            panic!("task is not valid: {:?}", e);
+        }
+        if let Some(kv_wb) = task.kv_wb.take() {
+            self.kv_wb.merge(kv_wb);
+        }
+        if let Some(raft_wb) = task.raft_wb.take() {
+            self.raft_wb.merge(raft_wb);
+        }
+
+        let entries = std::mem::take(&mut task.entries);
+        self.raft_wb.append(task.region_id, entries).unwrap();
+        if let Some((from, to)) = task.cut_logs {
+            self.raft_wb.cut_logs(task.region_id, from, to);
+        }
+        if let Some(raft_state) = task.raft_state.take() {
+            if self
+                .raft_states
+                .insert(task.region_id, raft_state)
+                .is_none()
+            {
+                self.state_size += std::mem::size_of::<RaftLocalState>();
+            }
+        }
+        self.readies
```

Comment:
What if different writes from different peers are batched together?

Reply:
It should not happen because the peer must be destroyed after all writes are done.

Reply:
How about checking the returned value to ensure this?